cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/AuthBinding.java

```diff
@@ -103,11 +107,111 @@
+
+  AuthBinding(String sentrySite, final String instanceName, final String sentryAdminGroup,
+              int cacheTtlSecs, int cacheMaxEntries) {
     this.authConf = initAuthzConf(sentrySite);
     this.instanceName = instanceName;
     this.authProvider = createAuthProvider();
     this.sentryAdminGroup = sentryAdminGroup;
+
+    groupCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<Principal, Set<String>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Set<String> load(Principal principal) throws Exception {
+          LOG.trace("Group cache miss for principal {}", principal);
+          return getGroups(principal);
+        }
+      });
+
+    roleCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<String, Set<Role>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Set<Role> load(final String group) throws Exception {
+          LOG.trace("Role cache miss for group {}", group);
+          return getRoles(group);
+        }
+      });
+
+    entityActionsCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<Role, Map<EntityId, Set<Action>>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Map<EntityId, Set<Action>> load(final Role role) throws Exception {
+          LOG.trace("Privilege cache miss for role {}", role);
+          return fetchEntityActions(role);
+        }
+      });
+  }
+
+  private Set<String> getGroups(Principal principal) {
+    return authProvider.getGroupMapping().getGroups(principal.getName());
+  }
+
+  private Set<Role> getRoles(final String group) throws Exception {
+    Set<TSentryRole> tSentryRoles = execute(new Command<Set<TSentryRole>>() {
+      @Override
+      public Set<TSentryRole> run(SentryGenericServiceClient client) throws Exception {
+        return client.listRolesByGroupName(sentryAdminGroup, group, COMPONENT_NAME);
+      }
+    });
+    Set<Role> roles = new HashSet<>();
+    for (TSentryRole tSentryRole : tSentryRoles) {
+      roles.add(new Role(tSentryRole.getRoleName()));
+    }
+    return roles;
+  }
+
+  private Map<EntityId, Set<Action>> fetchEntityActions(final Role role) throws Exception {
+    Set<TSentryPrivilege> sentryPrivileges = execute(new Command<Set<TSentryPrivilege>>() {
+      @Override
+      public Set<TSentryPrivilege> run(SentryGenericServiceClient client) throws Exception {
+        return client.listPrivilegesByRoleName(sentryAdminGroup, role.getName(), COMPONENT_NAME, instanceName);
+      }
+    });
+    Set<Privilege> privileges = toPrivileges(sentryPrivileges);
+
+    if (privileges == null) {
```

Comment:
This can never be bull since we do Set<Privilege> privileges = new HashSet<>(); in toPrivileges. I think your intention was to check for sentryPrivileges ? Although, I think that will also be empty list if no privilege exists rather than null.